encoding/dot/decode.go

```diff
@@ -0,0 +1,223 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dot
+
+import (
+	"fmt"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/encoding/dot/dotparser/ast"
+)
+
+// Builder is a graph that can have user-defined nodes and edges added.
+type Builder interface {
+	graph.Graph
+	graph.Builder
+	// NewNode adds a new node with a unique node ID to the graph.
+	NewNode() graph.Node
+	// NewEdge adds a new edge from the source to the destination node to the
+	// graph, or returns the existing edge if already present.
+	NewEdge(from, to graph.Node) graph.Edge
+}
+
+// UnmarshalerAttr is the interface implemented by objects that can unmarshal a
+// DOT attribute description of themselves.
+type UnmarshalerAttr interface {
+	// UnmarshalDOTAttr decodes a single DOT attribute.
+	UnmarshalDOTAttr(attr Attribute) error
+}
+
+// Copy copies the nodes and edges from the Graphviz AST source graph to the
+// destination graph. Edge direction is maintained if present.
+func Copy(dst Builder, src *ast.Graph) (err error) {
```

Comment:
Can I suggest then that you have `Unmarshal(src []byte, v interface{}) error` (or `src io.Reader`) and then type switch on `v` so that if it is `*[]dot.Builder` you unmarshal all the graphs in `src` into the elements of `v`, and if it is `dot.Builder` you unmarshal the single graph in `src` into it and return an error if there is more that one graph described by `src`. The use of a `[]dot.Builder` is a bit unpleasant here, but I expect it would be the minority case.